jmolecules-events/src/main/java/org/jmolecules/event/annotation/DomainEventHandler.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jmolecules.event.annotation;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Identifies a domain event handler, i.e. logic to process a {@link DomainEvent}.
+ *
+ * @author Christian Stettler
+ * @author Henning Schwentner
+ * @author Stephan Pirnbaum
+ * @author Martin Schimak
+ * @author Oliver Drotbohm
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD)
+@Documented
+public @interface DomainEventHandler {
+
+    /**
+     * Optional identification of the domain event handled by this handler.
+     * This information may be used for easier linkage between event and handler
+     * by external tools and refers to the type of the domain event.
+     */
+    String handles() default "";
```

Comment:
Maybe good to mention in the javadoc that leaving this type default will assume that the parameter(s) defines the type of event being handled.

Reply:
Indeed. One question though: Since we decided to split the identification of an event into `namespace` and `name`, what do you think should be the value to put here?

If we assume that the `name` without `namespace` is unique, this would be fine. However, if we have a handler that consumes multiple events from a namespace, this wouldn't match. So either the format needs to be something like: `<namespace>:<name>`, where we allow `*` for both or we add two fields (handlesNamespace, handlesName).

Reply:
Good question. @abuijze – Do you have input for that?

Reply:
I provided a proposal.

Reply:
Axon Framework currently uses a `Class<?>` field on the `@EventHandler` annotation to signal the type of event being handled by that handler. This essentially serves the purpose of an additional filter, to check if the event provided to the handling component actually matches that type.

Mirroring that with what's done here, I feel this reflects the `name`. The `namespace` to me resembles the (bounded) context in which the event has been published. Axon currently resolves the context on a message bus level, as the context in most scenarios belongs to the entire application. There is thus now no means in Axon to, on a handling annotation, define you would want events from a different context/namespace.

That would thus mean if Axon would employ these annotations, that `namespace` is fixed on a different level for the time being, and that the `name` is the only field which impacts "what" can be handled as defined ith the `handles` field on this annotation.

Massive if in my description here is that `namespace` == `context`, of which I am not entirely sure.
So, @abuijze, do you feel I am making a wrong assumption from Axon's perspective?
Similarly, @odrotbohm / @StephanPirnbaum, am I making a wrong assumption from JMolecules' perspective?

Reply:
Hi all, sorry for being late to the game, again. I still need to tune my github notifications so they are actually valuable...

I would argue that an event always has a name and a namespace. In that sense, they aren't optional. However, in the scope of the annotation they could be optional, where omission of the value means that a library is expected to inspect the signature of the annotated method to derive the values from the parameters.

In Axon, the first parameter is typically (at least by default) the actual event (or command, or query) to handle. The default namespace would be the package name of the declared event. The default name is the simple class name of that event.

If we have a handler that consumes multiple events, we should at least be able to specify more than one name/namespace combination. Maybe make the annotation repeatable? Since the name is most likely to be the differentiator here (though not guaranteed), using a "*" as a special value here could be useful. I don't think having the namespace of "*" makes any sense unless the name is also "*". For example, "OrderCreatedEvent" doesn't have any meaning unless you know in which context the word "order" is used. It could be an Order from Sales, but also an order from the purchasing department, which are two completely separate things, and most likely not even compatible...

Which brings be to the last question: `namespace` == `context`. I don't think they are `==`. Similar at best. Different contexts will certainly use different namespaces, but I can imagine that within a certain context, you may have more than one namespace.

Reply:
Okay, I think we're heading towards the finish line for this. Except for the polishing suggestions below, I think all additional concerns can be answered with non-breaking changes going forward.

> If we have a handler that consumes multiple events…

Can you elaborate how that'd work in a method declaration? Or would you use `@CommandHandler` on a type to e.g. pre-filter on the namespace and then just narrow down the handling on the method level? Feel free to point me to some Axon doc, if that is explained somewhere.

Here are the suggestions mentioned above:

* Augment the Javadoc for `@DomainEvent.namespace` and `….name` with "default… of the annotated type." Same for  the `@Command` annotation.
* Drop the `handles…` prefix from the annotation attributes in the `…Handler` annotations as they duplicate the prefix of the annotation. This'd also lead to symmetry between the annotation attributes of the event  / command and handler annotations.

If you think that's fine (esp. @StephanPirnbaum), please reply with a brief upvote. I can then take it from there.

Reply:
Axon Framework currently allows handling several types of an Event through the implementation of the Event.
So, if you'd have a `Loan` aggregate publishing 10 events, you could have an abstract/interface `LoanEvent` which is implemented by those 10 events.

In that case, Axon Framework allows you to specify an `@EventHandler` annotated method which automatically reacts to all those events. You can thus have an `@EventHandler` handling `Object` to receive *all* events provided to the handling class. By the way, this behaviour is *only* in place for Events. For Commands, an exact handler is simply required (as who's otherwise in charge of handling that operation), similarly for Queries (as what would otherwise be returned).

At the moment we are thinking of expanding this idea to serialized formats as well so that a user which does not know the concrete implementation (essentially the contract) to still be able to handle "everything".


Rounding this up, when @abuijze talks about "If we have a handler that consumes multiple events…" I feel he's talking about what this project will define. Axon Framework currently does not allow you to specify several events in the parameters of an `@EventHandler` annotated method, or through the annotation. Supporting something like that isn't overly strange if you ask me though.